File: 3. Behavioral Design Patterns/Visitor/Python/VisitorExample.py
# Visitor Pattern Example in Python
# Document structure with different element types processed by visitors

from abc import ABC, abstractmethod

class DocumentVisitor(ABC):
    @abstractmethod
    def visit_paragraph(self, paragraph):
        pass
    
    @abstractmethod
    def visit_image(self, image):
        pass
    
    @abstractmethod
    def visit_table(self, table):
        pass

class DocumentElement(ABC):
    @abstractmethod
    def accept(self, visitor):
        pass

class Paragraph(DocumentElement):
    def __init__(self, text):
        self.text = text
    
    def accept(self, visitor):
        visitor.visit_paragraph(self)

class Image(DocumentElement):
    def __init__(self, filename):
        self.filename = filename
    
    def accept(self, visitor):
        visitor.visit_image(self)

class Table(DocumentElement):
    def __init__(self, rows, columns):
        self.rows = rows
        self.columns = columns
    
    def accept(self, visitor):
        visitor.visit_table(self)

class WordCountVisitor(DocumentVisitor):
    def __init__(self):
        self.word_count = 0
    
    def visit_paragraph(self, paragraph):
        self.word_count += len(paragraph.text.split())
    
    def visit_image(self, image):
        # images don't contribute to word count
        pass
    
    def visit_table(self, table):
        self.word_count += table.rows * table.columns  # assume 1 word per cell

class PrintVisitor(DocumentVisitor):
    def visit_paragraph(self, paragraph):
        print(f"Paragraph: {paragraph.text}")
    
    def visit_image(self, image):
        print(f"Image: {image.filename}")
    
    def visit_table(self, table):
        print(f"Table: {table.rows}x{table.columns}")

def main():
    elements = [
        Paragraph("Hello world from paragraph one."),
        Image("photo.jpg"),
        Table(3, 4),
        Paragraph("This is another paragraph.")
    ]
    
    word_counter = WordCountVisitor()
    printer = PrintVisitor()
    
    for element in elements:
        element.accept(word_counter)
        element.accept(printer)
    
    print(f"Total word count: {word_counter.word_count}")

if __name__ == "__main__":
    main()

File: 3. Behavioral Design Patterns/Visitor/Typescript/VisitorExample.ts
// Visitor Pattern Example in TypeScript
// Document structure with different element types processed by visitors

interface DocumentVisitor {
  visitParagraph(paragraph: Paragraph): void
  visitImage(image: Image): void
  visitTable(table: Table): void
}

interface DocumentElement {
  accept(visitor: DocumentVisitor): void
}

class Paragraph implements DocumentElement {
  constructor(public text: string) {}
  accept(visitor: DocumentVisitor): void { visitor.visitParagraph(this) }
}

class Image implements DocumentElement {
  constructor(public filename: string) {}
  accept(visitor: DocumentVisitor): void { visitor.visitImage(this) }
}

class Table implements DocumentElement {
  constructor(public rows: number, public columns: number) {}
  accept(visitor: DocumentVisitor): void { visitor.visitTable(this) }
}

class WordCountVisitor implements DocumentVisitor {
  private _wordCount = 0
  get wordCount(): number { return this._wordCount }
  visitParagraph(paragraph: Paragraph): void { this._wordCount += paragraph.text.split(/\s+/).length }
  visitImage(image: Image): void { /* images don't contribute to word count */ }
  visitTable(table: Table): void { this._wordCount += table.rows * table.columns /* assume 1 word per cell */ }
}

class PrintVisitor implements DocumentVisitor {
  visitParagraph(paragraph: Paragraph): void { console.log(`Paragraph: ${paragraph.text}`) }
  visitImage(image: Image): void { console.log(`Image: ${image.filename}`) }
  visitTable(table: Table): void { console.log(`Table: ${table.rows}x${table.columns}`) }
}

;(function main() {
  const elements: DocumentElement[] = [
    new Paragraph('Hello world from paragraph one.'),
    new Image('photo.jpg'),
    new Table(3, 4),
    new Paragraph('This is another paragraph.')
  ]

  const wordCounter = new WordCountVisitor()
  const printer = new PrintVisitor()

  for (const element of elements) {
    element.accept(wordCounter)
    element.accept(printer)
  }

  console.log(`Total word count: ${wordCounter.wordCount}`)
})()

export {}
